test/model/autocomplete_test.dart

```diff
@@ -178,8 +184,106 @@ void main() {
+    doTest('#abc', null); // (no cursor)
+
+    doTest('~#^', channelLink(''));
+    doTest('~##^', channelLink('#'));
+    doTest('~#abc^', channelLink('abc'));
+    doTest('~#abc ^', channelLink('abc '));
+    doTest('~#abc def^', channelLink('abc def'));
+
+    // Accept space before channel link syntax.
+    doTest(' ~#abc^', channelLink('abc'));
+    doTest('xyz ~#abc^', channelLink('abc'));
+
+    // Accept punctuations before channel link syntax.
+    doTest(':~#abc^', channelLink('abc'));
+    doTest('!~#abc^', channelLink('abc'));
+    doTest(',~#abc^', channelLink('abc'));
+    doTest('.~#abc^', channelLink('abc'));
+    doTest('(~#abc^', channelLink('abc')); doTest(')~#abc^', channelLink('abc'));
+    doTest('{~#abc^', channelLink('abc')); doTest('}~#abc^', channelLink('abc'));
+    doTest('[~#abc^', channelLink('abc')); doTest(']~#abc^', channelLink('abc'));
+    doTest('“~#abc^', channelLink('abc')); doTest('”~#abc^', channelLink('abc'));
+    doTest('«~#abc^', channelLink('abc')); doTest('»~#abc^', channelLink('abc'));
+    // … and other punctuations except '#' and '@':
+    doTest('~##abc^', channelLink('#abc'));
+    doTest('~@#abc^', mention('#abc'));
+
+    // Avoid other characters before channel link syntax.
+    doTest('+#abc^', null);
+    doTest('=#abc^', null);
+    doTest('\$#abc^', null);
+    doTest('zulip/zulip-flutter#124^', null);
+    doTest('XYZ#abc^', null);
+    doTest('xyz#abc^', null);
+    // … but
+    doTest('~#xyz#abc^', channelLink('xyz#abc'));
+
+    // Avoid leading space character in query.
+    doTest('# ^', null);
+    doTest('# abc^', null);
+
+    // Avoid line-break characters in query.
+    doTest('#\n^', null); doTest('#a\n^', null); doTest('#\na^', null); doTest('#a\nb^', null);
+    doTest('#\r^', null); doTest('#a\r^', null); doTest('#\ra^', null); doTest('#a\rb^', null);
+    doTest('#\r\n^', null); doTest('#a\r\n^', null); doTest('#\r\na^', null); doTest('#a\r\nb^', null);
+
+    // Allow all other sorts of characters in query.
+    doTest('~#\u0000^', channelLink('\u0000')); // control
+    doTest('~#\u061C^', channelLink('\u061C')); // format character
+    doTest('~#\u0600^', channelLink('\u0600')); // format
+    doTest('~#\uD834^', channelLink('\uD834')); // leading surrogate
+    doTest('~#`^', channelLink('`'));   doTest('~#a`b^', channelLink('a`b'));
+    doTest('~#\\^', channelLink('\\')); doTest('~#a\\b^', channelLink('a\\b'));
+    doTest('~#"^', channelLink('"'));   doTest('~#a"b^', channelLink('a"b'));
+    doTest('~#>^', channelLink('>'));   doTest('~#a>b^', channelLink('a>b'));
+    doTest('~#&^', channelLink('&'));   doTest('~#a&b^', channelLink('a&b'));
+    doTest('~#_^', channelLink('_'));   doTest('~#a_b^', channelLink('a_b'));
+    doTest('~#*^', channelLink('*'));   doTest('~#a*b^', channelLink('a*b'));
+
+    // Two leading stars ('**') in the query are omitted.
+    doTest('~#**^', channelLink(''));
+    doTest('~#**abc^', channelLink('abc'));
```

Comment:
Similarly, the description of these test cases is focused on the logic in the implementation, and I'd like to instead focus on the user-facing goal which we want this code to accomplish and which this test verifies it accomplishes. Like this:
```suggestion
    // Accept syntax like "#**foo" (as from the user finishing an autocomplete
    // and then hitting backspace to edit it), but leave the "**" out of the query.
    doTest('~#**^', channelLink(''));
    doTest('~#**abc^', channelLink('abc'));
```